File: web/src/ui/root/dictionaryText.ts
/**
 * Pure helpers for the Root Viewer Dictionary section.
 * Mirror of Android `DictionaryText.kt` — keep the two in step.
 */
import type { DictionaryEntry, DictionarySenseGroup } from '../../data/dictionary'

/** Senses shown under Lemma before the reader asks for the rest. */
export const DICTIONARY_PREVIEW_SENSES = 4

export function wiktionaryPosForQac(qacPos: string): string | null {
  switch (qacPos.toUpperCase()) {
    case 'V':
    case 'VERB':
      return 'verb'
    case 'N':
    case 'NOUN':
      return 'noun'
    case 'ADJ':
    case 'ADJECTIVE':
      return 'adj'
    case 'ADV':
    case 'ADVERB':
      return 'adv'
    case 'P':
    case 'PREP':
    case 'PREPOSITION':
      return 'prep'
    case 'CONJ':
    case 'CONJUNCTION':
      return 'conj'
    case 'PRON':
    case 'PRONOUN':
      return 'pron'
    case 'PART':
    case 'PARTICLE':
      return 'particle'
    case 'INL':
    case 'INTERJECTION':
      return 'intj'
    default:
      return null
  }
}

export function orderedDictionaryGroups(
  groups: DictionarySenseGroup[],
  qacPos: string,
): DictionarySenseGroup[] {
  const preferred = wiktionaryPosForQac(qacPos)
  if (!preferred) return groups
  const match = groups.filter((g) => g.pos.toLowerCase() === preferred)
  const rest = groups.filter((g) => g.pos.toLowerCase() !== preferred)
  return [...match, ...rest]
}

export function dictionaryPosLabel(pos: string): string {
  switch (pos.toLowerCase()) {
    case 'verb':
      return 'Verb'
    case 'noun':
      return 'Noun'
    case 'adj':
    case 'adjective':
      return 'Adjective'
    case 'adv':
    case 'adverb':
      return 'Adverb'
    case 'prep':
    case 'preposition':
      return 'Preposition'
    case 'conj':
    case 'conjunction':
      return 'Conjunction'
    case 'pron':
    case 'pronoun':
      return 'Pronoun'
    case 'particle':
      return 'Particle'
    case 'intj':
    case 'interjection':
      return 'Interjection'
    case 'name':
    case 'proper noun':
    case 'propn':
      return 'Proper noun'
    default:
      return pos ? pos[0]!.toUpperCase() + pos.slice(1) : pos
  }
}

/** Flat gloss rows: optional POS label on the first gloss of each group. */
export function dictionaryGlosses(
  entry: DictionaryEntry,
  qacPos: string,
): Array<{ pos: string | null; gloss: string }> {
  const out: Array<{ pos: string | null; gloss: string }> = []
  for (const group of orderedDictionaryGroups(entry.groups, qacPos)) {
    const label = dictionaryPosLabel(group.pos)
    group.glosses.forEach((gloss, index) => {
      out.push({ pos: index === 0 ? label : null, gloss })
    })
  }
  return out
}

export function dictionaryNeedsExpand(glossCount: number): boolean {
  return glossCount > DICTIONARY_PREVIEW_SENSES
}

export function wiktionaryArabicUrl(word: string): string {
  const encoded = encodeURIComponent(word).replace(/%20/g, '_')
  return `https://en.wiktionary.org/wiki/${encoded}#Arabic`
}

File: app/src/main/java/com/beautifulquran/ui/rootviewer/DictionaryText.kt
package com.beautifulquran.ui.rootviewer

import com.beautifulquran.data.model.DictionaryEntry
import com.beautifulquran.data.model.DictionarySenseGroup
import java.net.URLEncoder

/** Senses shown under Lemma before the reader asks for the rest. */
internal const val DICTIONARY_PREVIEW_SENSES = 4

/**
 * Orders Wiktionary POS groups so the open word's QAC part of speech comes
 * first; everything else keeps source order after that.
 */
internal fun orderedDictionaryGroups(
    groups: List<DictionarySenseGroup>,
    qacPos: String,
): List<DictionarySenseGroup> {
    val preferred = wiktionaryPosForQac(qacPos) ?: return groups
    val (match, rest) = groups.partition { it.pos.equals(preferred, ignoreCase = true) }
    return match + rest
}

internal fun wiktionaryPosForQac(qacPos: String): String? = when (qacPos.uppercase()) {
    "V", "VERB" -> "verb"
    "N", "NOUN" -> "noun"
    "ADJ", "ADJECTIVE" -> "adj"
    "ADV", "ADVERB" -> "adv"
    "P", "PREP", "PREPOSITION" -> "prep"
    "CONJ", "CONJUNCTION" -> "conj"
    "PRON", "PRONOUN" -> "pron"
    "PART", "PARTICLE" -> "particle"
    "INL", "INTERJECTION" -> "intj"
    else -> null
}

internal fun dictionaryPosLabel(pos: String): String = when (pos.lowercase()) {
    "verb" -> "Verb"
    "noun" -> "Noun"
    "adj", "adjective" -> "Adjective"
    "adv", "adverb" -> "Adverb"
    "prep", "preposition" -> "Preposition"
    "conj", "conjunction" -> "Conjunction"
    "pron", "pronoun" -> "Pronoun"
    "particle" -> "Particle"
    "intj", "interjection" -> "Interjection"
    "name", "proper noun", "propn" -> "Proper noun"
    else -> pos.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
}

/** Flat gloss list in display order (preferred POS first). */
internal fun dictionaryGlosses(
    entry: DictionaryEntry,
    qacPos: String,
): List<Pair<String?, String>> {
    val out = ArrayList<Pair<String?, String>>()
    for (group in orderedDictionaryGroups(entry.groups, qacPos)) {
        val label = dictionaryPosLabel(group.pos)
        group.glosses.forEachIndexed { index, gloss ->
            out += (if (index == 0) label else null) to gloss
        }
    }
    return out
}

internal fun dictionaryNeedsExpand(glossCount: Int): Boolean =
    glossCount > DICTIONARY_PREVIEW_SENSES

/** Wiktionary page for the matched Arabic headword. */
internal fun wiktionaryArabicUrl(word: String): String {
    val encoded = URLEncoder.encode(word, Charsets.UTF_8.name()).replace("+", "_")
    return "https://en.wiktionary.org/wiki/$encoded#Arabic"
}
